### PR TITLE
Adding a `text_to_instance` method on DatasetReader

### LINKED ISSUE #190: Add a `text_to_instance` method on `DatasetReader`
The `Predictors` currently have to pull out the tokenizer and the token indexers from the `DatasetReader` and recreate what the `DatasetReader` does internally for every instance.  This means that if we want to change (or add parameters to) what the `DatasetReader` does, we have to change the `Predictors` to match.  Instead, we should just add a `text_to_instance` method on the `DatasetReader` itself, so that the `Predictor` can just keep around the `DatasetReader`, and pass off all processing to it.

### PR DESCRIPTION
**Pull request by matt-gardner**
This is so that the `Predictors`, or anything else that wants to make
predictions given a trained model, has an easier time of processing the
data in the same way the trained model did - all you need is the
instantiated `DatasetReader`, and not any private members pulled out
from that class.

Changing this on the `DatasetReaders`, though, meant changing the
`Predictors` themselves.  And that meant unifying the API around making
model predictions from single instance inputs.  I was able to remove
all of the `tag` and `predict_` methods on individual models, replacing
them with a single `Model.forward_on_instance` method.  I _think_ the
result is a lot cleaner, but I could have missed something important
when making this change - it was a whole bunch of pulling things apart
and putting them back together.

Fixes #190.

### CODE REVIEW

File: allennlp/service/predictors/semantic_role_labeler.py [hunk 1]
```diff
@@ -61,16 +69,20 @@ def predict_json(self, inputs: JsonDict) -> JsonDict:
...
         results = {"words": words, "verbs": []}  # type: JsonDict
-        text = TextField(words, token_indexers=self.token_indexers)
         for i, word in enumerate(spacy_doc):
             if word.pos_ == "VERB":
+                verb = word.text
                 verb_labels = [0 for _ in words]
                 verb_labels[i] = 1
-                verb_indicator = SequenceLabelField(verb_labels, text)
-                output = self.model.tag(text, verb_indicator)
+                instance = self._dataset_reader.text_to_instance(words, verb_labels)
```
Comment (DeNeutoy) [inline allennlp/service/predictors/semantic_role_labeler.py]: This hides the viterbi inference logic from the user, which i'm not sure is too great - this change has made it more difficult to use the model in a python script unless you use the `Predictor` (which is of course possible) - I had kind of thought that these classes were meant to be JSON -> JSON wrappers around models with complete functionality. If that's not how you thought of the `Predictor` wrappers, that's fine, but to me, this seems like a blurring of roles.

File: allennlp/service/predictors/simple_tagger.py [hunk 2]
```diff
@@ -8,15 +15,26 @@ class SimpleTaggerPredictor(Predictor):
...
 
-        return sanitize(self.model.tag(tokens))
+    @overrides
+    def _predictions_to_json(self, model_outputs: Dict[str, Any]) -> JsonDict:
+        predictions = model_outputs['class_probabilities']
+        argmax_indices = numpy.argmax(predictions, axis=-1)
+        tags = [self._model.vocab.get_token_from_index(x, namespace="labels")
+                for x in argmax_indices]
+        model_outputs['tags'] = tags
+        return super(SimpleTaggerPredictor, self)._predictions_to_json(model_outputs)
```
Comment (DeNeutoy) [inline allennlp/service/predictors/simple_tagger.py]: I would argue this is more complex than just having `return sanitize(model_outputs)` in each `Predictor`.

File: allennlp/service/predictors/semantic_role_labeler.py [hunk 3]
```diff
@@ -61,16 +69,20 @@ def predict_json(self, inputs: JsonDict) -> JsonDict:
...
+                output = self._model.forward_on_instance(instance)
+                transition_matrix = self._model.get_viterbi_pairwise_potentials()
+
+                predictions = torch.from_numpy(output['class_probabilities'])
+                max_likelihood_sequence, _ = viterbi_decode(predictions, transition_matrix)
+                tags = [self._model.vocab.get_token_from_index(x, namespace="labels")
+                        for x in max_likelihood_sequence]
```
Comment (DeNeutoy) [inline allennlp/service/predictors/semantic_role_labeler.py]: The fact that below, you have kept `return sanitize(results)` seems like an argument for removing `_predictions_to_json`.

File: allennlp/data/dataset_readers/semantic_role_labeling.py [hunk 4]
```diff
@@ -266,6 +266,24 @@ def read(self, file_path: str):
                                      "Is the path correct?".format(file_path))
         return Dataset(instances)
 
+    def text_to_instance(self,  # type: ignore
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/dataset_readers/semantic_role_labeling.py]: Done.

File: allennlp/service/predictors/predictor.py [hunk 5]
```diff
@@ -21,16 +19,28 @@ class Predictor(Registrable):
...
+        self._model = model
+        self._dataset_reader = dataset_reader
 
     def predict_json(self, inputs: JsonDict) -> JsonDict:
-        raise NotImplementedError()
+        instance = self._json_to_instance(inputs)
+        outputs = self._model.forward_on_instance(instance)
+        return self._predictions_to_json(outputs)
+
+    def _json_to_instance(self, json: JsonDict) -> Instance:
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/service/predictors/predictor.py]: The reason was for inference that needs to happen after `model.forward`.  Seems like there are a few options here:

1. We do what I did here, which is to put the call to inference / decoding logic in the `Predictor`, in a method that happens after the call to `model.forward_on_instance`.  It really should be an overridable method, because not every model needs to do this inference / decoding, and how that logic works depends on the model and the predictor.
2. We put the inference / decoding logic into `model.forward_on_instance`.  Either this means that models that do some kind of decoding need to put that decoding into `model.forward()` itself, or they need to override `model.forward_on_instance` to do the decoding there.  Because this probably means we're adding it to `model.forward()`, you probably want some kind of flag that says whether to do inference, because you probably don't want to bother at training time, but you might for the validation set.
3. We add another method on `Model` called `decode`, or something, that takes the output of `model.forward_on_instance` (or `model.forward`, but one of these has pytorch variables and the other has numpy - doing the conversion inside of `decode` wouldn't be hard...), and adds a decoded field to the dictionary.  The default implementation in the base class would just return an un-modified dictionary.  Then we can just add a call to `model.decode()` in `Predictor.predict_json()`, and remove the `_predictions_to_json` stuff.

I like option 3, I think.  It gives a cleaner separation between the jobs of `model.forward()` (which computes potentials / probabilities and a loss) and a decoding step (which does some kind of max inference / marginalization / beam search / whatever, given the potentials), and it makes it easy to have the logic all live inside of `Model`, where it belongs, without making training slow by putting unnecessary decoding into `forward`.

What do you think?

[the remark this answers is not in the document]
Comment (DeNeutoy) [reply allennlp/service/predictors/predictor.py]: Yep 3 sounds good. We should also consider whether we want this function to be able to take a list of instances and run forward on all of them as a batch or something as well.

[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/service/predictors/predictor.py]: Yeah, that's a good point.  I'll start with just a single instance, and we can worry about expanding it to a batched case later when someone needs it.

File: allennlp/models/model.py [hunk 6]
```diff
@@ -85,6 +86,32 @@ def forward(self, input1, input2, targets=None):
...
+        Takes an :class:`~allennlp.data.instance.Instance`, which typically has raw text in it,
+        converts that text into arrays using this model's :class:`Vocabulary`, passes those arrays
+        through :func:`self.forward()`, and returns the result.  Before returning the result, we
+        convert any ``torch.autograd.Variables`` or ``torch.Tensors`` into numpy arrays and remove
+        the batch dimension.
+        """
+        # Hack to see what cuda device the model is on, so we know where to put these inputs.  For
+        # complicated models, or machines with multiple GPUs, this will not work.  I couldn't find
+        # a way to actually query what device a tensor / parameter is on.
+        cuda_device = 0 if next(self.parameters()).is_cuda else -1
```
Comment (joelgrus) [inline allennlp/models/model.py]: is it possible to have a pytorch module with no parameters? because wouldn't this crash in that case? or is that not worth worrying about

File: allennlp/service/predictors/predictor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (joelgrus) [reply allennlp/service/predictors/predictor.py]: +1 to having all the logic live inside the `Model`

File: allennlp/models/model.py (hunk 6 above)
Comment (matt-gardner) [reply allennlp/models/model.py]: Yes, it's possible.  But this is on `Model`, and if you don't have any parameters in your model, why are you using this library?

**Comment by matt-gardner** [comment]
@DeNeutoy, I moved the decoding logic back into `Model`, if you want to have another look.  If you don't want to have another look, I'll just merge it.
